pydantic_ai_slim/pydantic_ai/models/mistral.py

```diff
@@ -0,0 +1,575 @@
+            tool_choice = 'none'
+        else:
+            tool_choice = 'auto'
+
+        response = await self.client.chat.complete_async(
+            model=str(self.model_name),
+            messages=mistral_messages,
+            n=1,
+            tools=self._map_function_and_result_tools_definition(),
+            tool_choice=tool_choice,
+            stream=False,
+        )
+        assert response, 'A unexpected empty response.'
+        return response
+
+    async def _stream_completions_create(
+        self,
+        messages: list[Message],
+    ) -> MistralEventStreamAsync[MistralCompletionEvent]:
+        """Create a streaming completion request to the Mistral model."""
+        response: MistralEventStreamAsync[MistralCompletionEvent] | None = None
+        mistral_messages = [self._map_message(m) for m in messages]
+
+        if self.result_tools and self.function_tools or self.function_tools:
+            # Function Calling Mode
+            response = await self.client.chat.stream_async(
+                model=str(self.model_name),
+                messages=mistral_messages,
+                n=1,
+                tools=self._map_function_and_result_tools_definition(),
+                tool_choice='auto',  # Note WARN: If use 'any' it will be infinity loop with Model.
+            )
+
+        elif self.result_tools:
+            # Json Mode
+            schema: str | list[dict[str, Any]]
+            if len(self.result_tools) == 1:
+                schema = _generate_json_simple_schema(self.result_tools[0].parameters_json_schema)
+            else:
+                parameters_json_schemas = [tool.parameters_json_schema for tool in self.result_tools]
+                schema = _generate_jsom_simple_schemas(parameters_json_schemas)
+
+            mistral_messages.append(
+                MistralUserMessage(
+                    content=f"""Answer in JSON Object, respect this following format:\n```\n{schema}\n```\n"""
+                )
+            )
+            response = await self.client.chat.stream_async(
+                model=str(self.model_name),
+                messages=mistral_messages,
+                response_format={'type': 'json_object'},
+            )
+
+        else:
+            # Stream Mode
+            response = await self.client.chat.stream_async(
+                model=str(self.model_name), messages=mistral_messages, stream=True, n=1
+            )
+        assert response
+        return response
+
+    def _map_function_and_result_tools_definition(self) -> list[MistralTool] | None:
+        """Map function and result tools to MistralTool format.
+
+        Returns None if both function_tools and result_tools are empty.
+        """
+        tools = []
+
+        all_tools: list[ToolDefinition] = []
+        if self.function_tools:
+            all_tools.extend(self.function_tools)
+        if self.result_tools:
+            all_tools.extend(self.result_tools)
```

Comment:
with the simplification of types above, this becomes just 
```suggestion
        all_tools: list[ToolDefinition] = self.function_tools + self.result_tools
```